### PR TITLE
[SPARK-15724] Add benchmarks for performance over wide schemas

### PR DESCRIPTION
**Pull request by ericl**
## What changes were proposed in this pull request?

This adds microbenchmarks for tracking performance of queries over very wide or deeply nested DataFrames. It seems performance degrades when DataFrames get thousands of columns wide or hundreds of fields deep.
## How was this patch tested?

Current results included.

cc @rxin @JoshRosen

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/WideSchemaBenchmark.scala [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    16 nodes x 5 deep x 10000 rows (write)       100 /  126          1.0         999.5       0.6X
+    128 nodes x 8 deep x 1000 rows (read)         53 /   61          1.9         532.9       1.1X
+    128 nodes x 8 deep x 1000 rows (write)       530 /  656          0.2        5295.6       0.1X
+    512 nodes x 10 deep x 200 rows (read)         98 /  134          1.0         979.2       0.6X
+    512 nodes x 10 deep x 200 rows (write)      6698 / 7124          0.0       66977.0       0.0X
+    */
+  }
+
+  //
+  // The following benchmarks are for reference: the schema is not actually wide for them.
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/WideSchemaBenchmark.scala]: what does this mean?

**Comment by rxin** [comment]
For the ones that we actually are testing throughput (i.e. have more than 1 rows), can we generate temp files using parquet? Or at least add variant of that.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/WideSchemaBenchmark.scala [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    for (width <- widthsToTest) {
+      val selectExpr = (1 to width).map(i => s"id as a_$i")
+      benchmark.addCase(s"$width select expressions") { iter =>
+        sparkSession.range(1).toDF.selectExpr(selectExpr: _*)
+      }
+    }
+    benchmark.run()
+
+    /*
+    OpenJDK 64-Bit Server VM 1.8.0_66-internal-b17 on Linux 4.2.0-36-generic
```
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/WideSchemaBenchmark.scala]: These benckmarks are quite small In my experience C2 JIT compilation will only kick in after a few seconds. So I don't think these are running a full speed.

Comment (ericl) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/WideSchemaBenchmark.scala]: I addressed this in https://github.com/apache/spark/pull/13472: in practice the times didn't seem to change by that much (maybe 1.5x faster for the smaller cases).

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/WideSchemaBenchmark.scala (hunk 1 above)
Comment (ericl) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/WideSchemaBenchmark.scala]: Removed the comment, it's confusing.

**Comment by ericl** [comment]
@rxin updated with parquet variations

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/WideSchemaBenchmark.scala [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+  private lazy val sparkSession = SparkSession.builder
+    .master("local[1]")
+    .appName("microbenchmark")
+    .getOrCreate()
+
+  import sparkSession.implicits._
+
+  private var tmpFiles: List[File] = Nil
+
+  override def afterEach() {
```
Comment (JoshRosen) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/WideSchemaBenchmark.scala]: Nit: call `super.afterEach()` in a finally block.

Comment (ericl) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/WideSchemaBenchmark.scala]: Done

**Comment by JoshRosen** [comment]
LGTM. I'm going to merge this in for now so that we can divide-and-conquer on perf. improvement work. Let's deal with any additional feedback in followups.

I'm also going to merge this into `branch-2.0` so that PRs which update the benchmark results don't merge conflict as easily when cherry-picking them for 2.0.

**Pull request closed without merging**